presto-main/src/main/java/com/facebook/presto/operator/aggregation/Histogram.java

```diff
@@ -0,0 +1,175 @@
+        return new FunctionInfo(signature, getDescription(), aggregation);
+    }
+
+    private static InternalAggregationFunction generateAggregation(Type keyType, Type valueType)
+    {
+        DynamicClassLoader classLoader = new DynamicClassLoader(Histogram.class.getClassLoader());
+        List<Type> inputTypes = ImmutableList.of(keyType);
+        Type outputType = new MapType(keyType, valueType);
+        HistogramStateSerializer stateSerializer = new HistogramStateSerializer();
+        Type intermediateType = stateSerializer.getSerializedType();
+
+        AggregationMetadata metadata = new AggregationMetadata(
+                generateAggregationName(NAME, outputType, inputTypes),
+                createInputParameterMetadata(keyType),
+                INPUT_FUNCTION,
+                null,
+                null,
+                COMBINE_FUNCTION,
+                OUTPUT_FUNCTION,
+                HistogramState.class,
+                stateSerializer,
+                new HistogramStateFactory(keyType, EXPECTED_SIZE_FOR_HASHING),
+                outputType,
+                false);
+
+        GenericAccumulatorFactoryBinder factory = new AccumulatorCompiler().generateAccumulatorFactoryBinder(metadata, classLoader);
+        return new InternalAggregationFunction(NAME, inputTypes, intermediateType, outputType, true, false, factory);
+    }
+
+    private static List<ParameterMetadata> createInputParameterMetadata(Type keyType)
+    {
+        return ImmutableList.of(new ParameterMetadata(STATE),
+                new ParameterMetadata(BLOCK_INPUT_CHANNEL, keyType),
+                new ParameterMetadata(BLOCK_INDEX));
+    }
+
+    public static void input(HistogramState state, Block key, int position)
+    {
+        SingleChannelGroupByHash groupByHash = state.get();
+        if (groupByHash == null) {
+            groupByHash = new SingleChannelGroupByHash(state.getType(), state.getExpectedSizeForHashing());
+            state.set(groupByHash);
+        }
+
+        long startSize = groupByHash.getEstimatedSize();
+        try {
+            groupByHash.putIfAbsent(position, key, 1L);
+        }
+        catch (ExceededMemoryLimitException e) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("The result of histogram may not exceed %s", e.getMaxMemory()));
+        }
+        state.addMemoryUsage(groupByHash.getEstimatedSize() - startSize);
+    }
+
+    public static void combine(HistogramState state, HistogramState otherState)
+    {
+        if (state.get() != null && otherState.get() != null) {
+            Block valuesByGroupId = otherState.get().getValuesByGroupId();
+            LongBigArray countsByGroupId = otherState.get().getCountsByGroupId();
+            SingleChannelGroupByHash groupByHash = state.get();
+            long startSize = groupByHash.getEstimatedSize();
+            for (int i = 0; i < otherState.get().getGroupCount(); i++) {
```

Comment:
Instead of this loop refactor the code, so that `TypedHistogram` has a `addAll(TypedHistogram other)` method, which adds all the values and respective counts from other to the histogram